### PR TITLE
8330027: Identity hashes of archived objects must be based on a reproducible random seed

### PR DESCRIPTION
**Pull request by tstuefe**
This fixes an issue with CDS archive reproducibility that can happen under rare circumstances. See original JBS issue for details.

I had to manually resolve the hunk in src/hotspot/share/cds/heapShared.cpp because both

    8251330: Reorder CDS archived heap to speed up relocation
    8329431: Improve speed of writing CDS heap objects
    were missing.

I also had to change `CDSConfig::is_dumping_static_archive()` - which only exists in JDK 22 and later - to the technically equal `DumpSharedSpaces`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8330027](https://bugs.openjdk.org/browse/JDK-8330027) needs maintainer approval

### Issue
 * [JDK-8330027](https://bugs.openjdk.org/browse/JDK-8330027): Identity hashes of archived objects must be based on a reproducible random seed (**Bug** - P4 - Approved)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk21u-dev.git pull/562/head:pull/562` \
`$ git checkout pull/562`

Update a local copy of the PR: \
`$ git checkout pull/562` \
`$ git pull https://git.openjdk.org/jdk21u-dev.git pull/562/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 562`

View PR using the GUI difftool: \
`$ git pr show -t 562`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk21u-dev/pull/562.diff">https://git.openjdk.org/jdk21u-dev/pull/562.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk21u-dev/pull/562#issuecomment-2105569175)

### CODE REVIEW

File: src/hotspot/share/cds/heapShared.cpp [hunk 1]
```diff
@@ -277,11 +277,6 @@ bool HeapShared::archive_object(oop obj) {
     count_allocation(obj->size());
     ArchiveHeapWriter::add_source_obj(obj);
```
Comment (TheRealMDoerr) [inline src/hotspot/share/cds/heapShared.cpp:279]: Original change removes the empty line here.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM. I suggest removing the empty line, too.

**Comment by tstuefe** [comment]
Thanks @TheRealMDoerr !

**Review by TheRealMDoerr — APPROVED** [review]

**Comment by tstuefe** [comment]
/approval I would like to fix this in JDK 21, since it is important for our reproducible-build effort. 

Risk: A part of this patch is that we don't pro-actively create ihashes anymore during archival dump. I don't think that can cause any problem, nor does Ioi Lam who proposed that, but just to be aware - it may cause subtle changes in CDS dumps. The upstream patch has been in 23 for several weeks now, without any problems floating up.

**Comment by TheRealMDoerr** [comment]
@tstuefe: The approval command was wrong. Please take a look.

**Comment by tstuefe** [comment]
/approval request  I would like to fix this in JDK 21, since it is important for our reproducible-build effort.

**Pull request closed without merging**